File: src/main/java/me/tomasan7/jecnadesktop/repository/WebAttendancesRepository.java
package me.tomasan7.jecnadesktop.repository;

import me.tomasan7.jecnadesktop.data.Attendances;
import me.tomasan7.jecnadesktop.parser.parsers.HTMLAttendancesParser;
import me.tomasan7.jecnadesktop.web.JecnaWebClient;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

public class WebAttendancesRepository implements AttendancesRepository
{
	private static final String WEB_PATH = "/absence/passing-student";

	private final JecnaWebClient webClient;
	private final HTMLAttendancesParser attendancesParser = new HTMLAttendancesParser();

	public WebAttendancesRepository (JecnaWebClient webClient)
	{
		this.webClient = webClient;
	}

	@Override
	public Attendances queryAttendances ()
	{
		try
		{
			return attendancesParser.parse(webClient.query(WEB_PATH).get());
		}
		catch (ExecutionException | InterruptedException e)
		{
			throw new RuntimeException(e);
		}
	}

	@Override
	public CompletableFuture<Attendances> queryAttendancesAsync ()
	{
		return webClient.query(WEB_PATH)
						.thenApply(attendancesParser::parse);
	}
}

File: src/main/kotlin/me/tomasan7/jecnadesktop/repository/WebAttendancesRepository.kt
package me.tomasan7.jecnadesktop.repository

import io.ktor.http.*
import me.tomasan7.jecnadesktop.data.Attendances
import me.tomasan7.jecnadesktop.data.SchoolYear
import me.tomasan7.jecnadesktop.parser.parsers.HtmlAttendancesParser
import me.tomasan7.jecnadesktop.util.JecnaPeriodEncoder
import me.tomasan7.jecnadesktop.util.JecnaPeriodEncoder.jecnaEncode
import me.tomasan7.jecnadesktop.web.JecnaWebClient
import me.tomasan7.jecnadesktop.web.append

/**
 * Retrieves [Attendances] from the Ječná web.
 */
class WebAttendancesRepository(private val webClient: JecnaWebClient) : AttendancesRepository
{
    private val attendancesParser = HtmlAttendancesParser()

    override suspend fun queryAttendances() = attendancesParser.parse(webClient.queryStringBody(WEB_PATH))

    override suspend fun queryAttendances(schoolYear: SchoolYear, month: Int) =
        attendancesParser.parse(webClient.queryStringBody(WEB_PATH, Parameters.build {
            append(schoolYear.jecnaEncode())
            append(JecnaPeriodEncoder.encodeMonth(month))
        }))

    companion object
    {
        private const val WEB_PATH = "/absence/passing-student"
    }
}